### PR TITLE
IGNITE-21230 Custom serialization of catalog commands.

### PR DESCRIPTION
**Pull request by xtern**
https://issues.apache.org/jira/browse/IGNITE-21230

* Added `UpdateLogMarshaller` that implements external serialization for `VersionedUpdate`.
* For each type of catalog object, a separate `CatalogObjectSerializer` has been added.

### CODE REVIEW

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogTableDescriptor.java [hunk 1]
```diff
@@ -219,4 +221,53 @@ public void updateToken(long updateToken) {
...
         this.creationToken = this.creationToken == INITIAL_CAUSALITY_TOKEN ? updateToken : this.creationToken;
     }
+
+    /**
+     * Serializer for {@link CatalogTableDescriptor}.
+     */
+    private static class TableDescriptorSerializer implements CatalogObjectSerializer<CatalogTableDescriptor> {
+        @Override
+        public CatalogTableDescriptor readFrom(int version, IgniteDataInput input) throws IOException {
+            CatalogDescriptorBase header = CatalogObjectDescriptor.SERIALIZER.readFrom(version, input);
```
Comment (AMashenkov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogTableDescriptor.java:231]: Let's get rid of additional objects.
```suggestion
            int id = input.readInt()
            String name = input.readUtf();
            long input = readLong();
```

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogTableDescriptor.java [hunk 2]
```diff
@@ -219,4 +221,53 @@ public void updateToken(long updateToken) {
...
+                    colocationColumns,
+                    schemaVersions,
+                    header.updateToken(),
+                    creationToken
+            );
+        }
+
+        @Override
+        public void writeTo(CatalogTableDescriptor descriptor, int version, IgniteDataOutput output) throws IOException {
+            CatalogObjectDescriptor.SERIALIZER.writeTo(new CatalogDescriptorBase(descriptor), version, output);
```
Comment (AMashenkov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogTableDescriptor.java:261]:
```suggestion
            output.writeInt(id);
            writeStringCollection(...);
            output.writeLong(updateToken());
```

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogTableDescriptor.java [hunk 3]
```diff
@@ -219,4 +221,53 @@ public void updateToken(long updateToken) {
...
+            CatalogTableSchemaVersions.SERIALIZER.writeTo(descriptor.schemaVersions(), version, output);
+            writeList(descriptor.columns(), version, CatalogTableColumnDescriptor.SERIALIZER, output);
+
+            output.writeInt(descriptor.schemaId());
+            output.writeInt(descriptor.primaryKeyIndexId());
+            output.writeInt(descriptor.zoneId());
+            writeStringCollection(descriptor.primaryKeyColumns(), output);
+            writeStringCollection(descriptor.colocationColumns(), output);
```
Comment (AMashenkov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogTableDescriptor.java:269]: We can skip colocation columns if they match key columns.
I guess we pass the same collection (the same object) to constructor in that case.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogObjectDescriptor.java [hunk 4]
```diff
@@ -19,17 +19,22 @@
...
+import org.apache.ignite.internal.util.io.IgniteDataOutput;
 
 /**
  * Base class for catalog objects.
  * TODO: IGNITE-19082 Implement custom effective serialization instead.
  */
-public abstract class CatalogObjectDescriptor implements Serializable {
-    private static final long serialVersionUID = -6525237234280004860L;
+public abstract class CatalogObjectDescriptor {
+    public static CatalogDescriptorBaseSerializer SERIALIZER = new CatalogDescriptorBaseSerializer();
```
Comment (AMashenkov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogObjectDescriptor.java:36]: Let's avoid serialization for abstract classes.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/serialization/CatalogSerializationUtils.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        int size = in.readInt();
+
+        return readStringCollection(in, size, new ArrayList<>(size));
+    }
+
+    /** Reads set of strings. */
+    public static Set<String> readStringSet(DataInput in) throws IOException {
+        int size = in.readInt();
+
+        return readStringCollection(in, size, new HashSet<>());
```
Comment (AMashenkov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/serialization/CatalogSerializationUtils.java:77]:
```suggestion
        return readStringCollection(in, size, IgniteUtils.newHashSet(size));
```

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/serialization/CatalogSerializationUtils.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    public static void writeStringCollection(Collection<String> list, DataOutput out) throws IOException {
+        out.writeInt(list.size());
+
+        for (String item : list) {
+            out.writeUTF(item);
+        }
+    }
+
+    /** Reads list of strings. */
+    public static List<String> readStringList(DataInput in) throws IOException {
```
Comment (AMashenkov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/serialization/CatalogSerializationUtils.java:67]: You can have a single method instead.
```suggestion
    public static <T extends Collection<String>> T readCollection(DataInput in, IntFunction<T> factory) throws IOException {
```

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/serialization/UpdateLogMarshallerImpl.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+    @TestOnly
+    UpdateLogMarshallerImpl(int version, CatalogEntrySerializerProvider serializers) {
+        this.version = version;
+        this.serializers = serializers;
+    }
+
+    @Override
+    public byte[] marshall(VersionedUpdate update) {
+        IgniteUnsafeDataOutput output = new IgniteUnsafeDataOutput(32);
```
Comment (AMashenkov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/serialization/UpdateLogMarshallerImpl.java:58]:
```suggestion
        IgniteUnsafeDataOutput output = new IgniteUnsafeDataOutput(128);
```

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/serialization/UpdateLogMarshallerImpl.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+            for (UpdateEntry entry : update.entries()) {
+                output.writeShort(entry.typeId());
+
+                CatalogObjectSerializer<? super UpdateEntry> serializer = serializers.get(entry.typeId());
+                serializer.writeTo(entry, version, output);
+            }
+
+            output.writeInt(update.version());
+            output.writeLong(update.delayDurationMs());
+
```
Comment (AMashenkov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/serialization/UpdateLogMarshallerImpl.java:74]:
```suggestion
            output.writeInt(update.version());
            output.writeLong(update.delayDurationMs());
           
            output.writeInt(update.entries().size());
            for (UpdateEntry entry : update.entries()) {
                output.writeShort(entry.typeId());

                CatalogObjectSerializer<? super UpdateEntry> serializer = serializers.get(entry.typeId());
                serializer.writeTo(entry, version, output);
            }

```

File: modules/catalog/src/test/java/org/apache/ignite/internal/catalog/serialization/CatalogEntrySerializationCompatibilityTest.java [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+import org.apache.ignite.internal.util.io.IgniteDataInput;
+import org.apache.ignite.internal.util.io.IgniteDataOutput;
+import org.apache.ignite.lang.MarshallerException;
+import org.junit.jupiter.api.Test;
+
+/**
+ * Tests to verify catalog storage entries serialization compatibility.
+ */
+public class CatalogEntrySerializationCompatibilityTest {
+    private final CatalogEntrySerializerProvider serializerProvider = (ignore) -> TestUpdateEntrySerializer.INSTANCE;
```
Comment (AMashenkov) [inline modules/catalog/src/test/java/org/apache/ignite/internal/catalog/serialization/CatalogEntrySerializationCompatibilityTest.java:39]: You have a single serializer version here.
I think the test idea is to have at least 2 versions, "old" and "new" ones. Where "old" one supports version 1 only, and "new" one supports version 1 and 2.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogTableDescriptor.java (hunk 1 above)
Comment (xtern) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogTableDescriptor.java]: Removed atalogObjectDescriptor.SERIALIZER.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogTableDescriptor.java (hunk 2 above)
Comment (xtern) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogTableDescriptor.java]: Done

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogObjectDescriptor.java (hunk 4 above)
Comment (xtern) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogObjectDescriptor.java]: Done

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/serialization/CatalogSerializationUtils.java (hunk 5 above)
Comment (xtern) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/serialization/CatalogSerializationUtils.java]: didn't find this method, replaced with `new HashSet<>(capacity(size));`

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/serialization/CatalogSerializationUtils.java (hunk 6 above)
Comment (xtern) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/serialization/CatalogSerializationUtils.java]: Done, thanks

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/serialization/UpdateLogMarshallerImpl.java (hunk 7 above)
Comment (xtern) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/serialization/UpdateLogMarshallerImpl.java]: Done

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/serialization/UpdateLogMarshallerImpl.java (hunk 8 above)
Comment (xtern) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/serialization/UpdateLogMarshallerImpl.java]: Done

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/serialization/UpdateLogMarshallerImpl.java [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+/**
+ * Marshaller of update log entries that uses custom serializer.
+ */
+public class UpdateLogMarshallerImpl implements UpdateLogMarshaller {
+    /** Current data format version. */
+    private static final int PROTOCOL_VERSION = 1;
+
+    /** Required data format version. */
+    private final int version;
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/serialization/UpdateLogMarshallerImpl.java]: Done

File: modules/catalog/src/test/java/org/apache/ignite/internal/catalog/serialization/CatalogEntrySerializationCompatibilityTest.java (hunk 9 above)
Comment (xtern) [reply modules/catalog/src/test/java/org/apache/ignite/internal/catalog/serialization/CatalogEntrySerializationCompatibilityTest.java]: From my point of view it doesn't affect the test case but breaks it. But for clarity, I divided it into 3 different serializers.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogTableDescriptor.java (hunk 3 above)
Comment (xtern) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogTableDescriptor.java:274]: I'm not sure it's worth adding such optimization. As I understand we need to compare collections for equality here + store special marker (for example size = `-2`) that collection are equal. :thinking:

Comment (AMashenkov) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogTableDescriptor.java:274]: I think we can compare them by reference.

File: modules/catalog/src/test/java/org/apache/ignite/internal/catalog/serialization/CatalogEntrySerializationCompatibilityTest.java [hunk 11]
```diff
@@ -0,0 +1,429 @@
...
+import org.apache.ignite.internal.util.io.IgniteDataInput;
+import org.apache.ignite.internal.util.io.IgniteDataOutput;
+import org.apache.ignite.lang.MarshallerException;
+import org.junit.jupiter.api.Test;
+
+/**
+ * Tests to verify catalog storage entries serialization compatibility.
+ */
+public class CatalogEntrySerializationCompatibilityTest {
+    private final CatalogEntrySerializerProvider serializerProvider = (typeId) -> {
```
Comment (AMashenkov) [inline modules/catalog/src/test/java/org/apache/ignite/internal/catalog/serialization/CatalogEntrySerializationCompatibilityTest.java:39]: I think 2 versions is enough. The test must checks next cases:
* Marshaller of version X, must successfully marshal-unmarshal an entry.
* Marshaller of version 2, must successfully unmarshal an entry marshalled by marshaller version 1.
* Marshaller of version 1, must fails unmarshal an entry marshalled by marshaller version 2.

File: modules/catalog/src/test/java/org/apache/ignite/internal/catalog/serialization/CatalogEntrySerializationCompatibilityTest.java [hunk 12]
```diff
@@ -0,0 +1,429 @@
...
+            deserialized = marshallerV3.unmarshall(bytesV2);
+            assertThat(deserialized.entries(), equalTo(entries));
+        }
+    }
+
+    /**
+     * Checks whether the entry can be written in the previous format.
+     */
+    @Test
+    public void writePreviousVersionEntry() {
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/catalog/src/test/java/org/apache/ignite/internal/catalog/serialization/CatalogEntrySerializationCompatibilityTest.java:108]: The purpose of the test is to check whether, after changing the protocol version, we can work on the old version. We need to be able to write entries in the old format until the entire cluster can use the new format.

File: modules/catalog/src/test/java/org/apache/ignite/internal/catalog/serialization/CatalogEntrySerializationCompatibilityTest.java (hunk 11 above)
Comment (xtern) [reply modules/catalog/src/test/java/org/apache/ignite/internal/catalog/serialization/CatalogEntrySerializationCompatibilityTest.java:39]: Done, test simplified, removed 3rd version of entry and serializer.

```
Marshaller of version X, must successfully marshal-unmarshal an entry.
Marshaller of version 2, must successfully unmarshal an entry marshalled by marshaller version 1.
```
implemented in `readEntryInPreviousFormat`

```
Marshaller of version X, must successfully marshal-unmarshal an entry.
Marshaller of version 1, must fails unmarshal an entry marshalled by marshaller version 2.
```
implemented in `forwardCompatibilityIsNotSupported`

```
Serializer of version 2 can *write* entry in version 1 format
```
implemented in `writeEntryInPreviousFormat`

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogTableDescriptor.java (hunk 3 above)
Comment (xtern) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogTableDescriptor.java]: Done, thanks.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogColumnCollation.java [hunk 13]
```diff
@@ -70,4 +70,17 @@ public boolean asc() {
...
+
+    /** Writes collation to byte. */
+    static byte pack(CatalogColumnCollation collation) {
+        return  (byte) ((collation.asc() ? 1 : 0) | (collation.nullsFirst() ? 2 : 0));
+    }
+
+    /** Reads collation from byte. */
+    static CatalogColumnCollation unpack(byte collationBits) {
+        boolean asc = (collationBits & 1) == 1;
+        boolean nullsFirst = (collationBits & 2) == 2;
```
Comment (AMashenkov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogColumnCollation.java:82]: Let's convert magic numbers to constants.

Comment (xtern) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogColumnCollation.java]: Added constants for bit positions.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogIndexDescriptor.java [hunk 14]
```diff
@@ -66,8 +67,40 @@ public int creationCatalogVersion() {
...
     public String toString() {
         return S.toString(CatalogIndexDescriptor.class, this, super.toString());
     }
+
+    /** Catalog index descriptor type. */
+    public enum CatalogIndexDescriptorType {
+        HASH(0),
+        SORTED(1);
+
+        private final int id;
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogIndexDescriptor.java]: Done.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/serialization/UpdateLogMarshallerImpl.java [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+ *
+ *<p>Basic format description:
+ *<pre>
+ * (size) | description
+ * ------------------------------
+ *     2  | data format version ({@link #PROTOCOL_VERSION})
+ *     2  | entry type ({@link VersionedUpdate} or {@link SnapshotEntry})
+ *
+ * [fields of the corresponding event follow (for example, {@link VersionedUpdate}]
+ *
```
Comment (AMashenkov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/serialization/UpdateLogMarshallerImpl.java:51]:
```suggestion
 *     2  | entry type ({@link VersionedUpdate} or {@link SnapshotEntry})
 * <list> | fields of the corresponding event follow (for example, {@link VersionedUpdate}
 * </pre>
 * 
 * Versioned Update format description:
 * <pre>
```

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/serialization/UpdateLogMarshallerImpl.java [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+ *     4  | version
+ *     8  | delayDurationMs
+ *
+ * [list of entries follow]
+ *
+ *     4  | entries size
+ *     2  | entry type (see {@link MarshallableEntryType#id()})
+ *
+ * [entry fields follow]
```
Comment (AMashenkov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/storage/serialization/UpdateLogMarshallerImpl.java:62]:
```suggestion
 * <list> | list of {@link MarshallableEntry}
 * </pre>
 *
 * MarshallableEntry format:
 * <pre>
 *     4  | entries size
 *     2  | entry type (see {@link MarshallableEntryType#id()})
 *       | other entry fields.
```

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogColumnCollation.java [hunk 17]
```diff
@@ -27,6 +27,8 @@ public enum CatalogColumnCollation {
     DESC_NULLS_FIRST(false, true),
     DESC_NULLS_LAST(false, false);
 
+    private static final int ASC_BIT_POS = 0;
+    private static final int NULLS_FIRST_BIT_POS = 1;
```
Comment (AMashenkov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogColumnCollation.java:31]:
```suggestion
    private static final int ASC_BIT_POS = 1;
    private static final int NULLS_FIRST_BIT_POS = 1 << 1;
```

Comment (xtern) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogColumnCollation.java]: :thinking: :thinking:  reverted last changes and renamed to ASC_FLAG_**BITMASK**...

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogColumnCollation.java [hunk 18]
```diff
@@ -27,6 +27,8 @@ public enum CatalogColumnCollation {
     DESC_NULLS_FIRST(false, true),
     DESC_NULLS_LAST(false, false);
 
+    private static final int ASC_FLAG_BITMASK = 1;
+    private static final int NULLS_FIRST_FLAG_BITMASK = 1 << 2;
```
Comment (AMashenkov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogColumnCollation.java:31]:
```suggestion
    private static final int NULLS_FIRST_FLAG_BITMASK = 1 << 1;
```

**Review by AMashenkov — APPROVED** [review]

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogColumnCollation.java (hunk 18 above)
Comment (xtern) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogColumnCollation.java:31]: Thanks, missed it.

**Pull request merged**